pallets/pool-system/src/lib.rs

```diff
@@ -1239,15 +1240,37 @@ pub mod pallet {
 
 			let executed_amounts = epoch.tranches.fulfillment_cash_flows(solution)?;
 			let total_assets = epoch.nav.total(pool.reserve.total)?;
-			let tranche_ratios = epoch.tranches.combine_with_residual_top(
-				&executed_amounts,
-				|tranche, &(invest, redeem)| {
-					Ok(Perquintill::from_rational(
-						tranche.supply.ensure_add(invest)?.ensure_sub(redeem)?,
-						total_assets,
-					))
-				},
-			)?;
+			let mut tranche_ratios = epoch
+				.tranches
+				.non_residual_tranches()
+				.map(|tranches| {
+					tranches
+						.iter()
+						// NOTE: Reversing amounts, as residual amount is on top.
+						// NOTE: Iterator of executed amounts is one time larger than the
+						//       non_residual_tranche-iterator, but we anyways push all reamaining
+						//       ratio to the residual tranche.
+						.zip(executed_amounts.rev())
+						.map(|(tranche, &(invest, redeem))| {
+							Ok(Perquintill::from_rational(
+								tranche.supply.ensure_add(invest)?.ensure_sub(redeem)?,
+								total_assets,
+							))
+						})
+						.collect::<Result<Vec<_>, ArithmeticError>>()
+				})
```

Comment:
Can"t we simplify this by using existing `Tranches` API or does this not work because all API includes the residual tranche?

Reply:
it is already using tranches, but not the combine as I, as you said, need to exclude the residual tranche.

Reply:
Happy to add a cleaner solution if you have one at hand.

Reply:
See https://github.com/centrifuge/centrifuge-chain/pull/1856/commits/7886f40345db4e438b72ab5182b2e7fa2df1394f